### PR TITLE
PLT-1378 Initial version of emoji reactions

### PR DESCRIPTION
**Pull request by hmhealey**
Adds the infrastructure required for emoji reactions without the emoji picker which was causing problems. In addition to what was demoed/tested before 3.4, this reorganizes the emoji code to support some features needed by the emoji picker including adding support for emoji categories and proper aliases.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-1378

#### Checklist
- Added or updated unit tests (required for all new features)
- Added API documentation (https://github.com/mattermost/mattermost-api-reference/pull/14)
- All new/modified APIs include changes to the drivers
- Has UI changes
- Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates

### CODE REVIEW

**Comment by moonwolf-github** [comment]
:+1:  (sorry i couldn't help myself)

**Comment by lfbrock** [comment]
@hmhealey is there a good way to put this behind a feature flag?

I found one new bug, repro steps are: 

1. Click reply on a post with emojis
2. Click reply on a different post with emojis

Observed: The post text updates, but the emojis from the old post still show. 
Expected: The emojis should be updated as well, so the correct emojis show for the post. 


There's also still the known issue of the post wrapping for a lot of emojis, but we'll fix this later. 
screenshot 2016-11-16 10 58 29

**Comment by hmhealey** [comment]
Updated to fix that issue. I'll restart the test server once it has rebuilt

**Comment by lfbrock** [comment]
+1 - I think the emoji wrapping problem is still happening, but otherwise looks good.

File: store/sql_reaction_store.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+func saveReactionAndUpdatePost(transaction *gorp.Transaction, reaction *model.Reaction) error {
+	if count, err := transaction.SelectInt(
+		`SELECT
+			COUNT(0)
+		FROM
+			Reactions
+		WHERE
+			PostId = :PostId AND
+			UserId = :UserId AND
+			EmojiName = :EmojiName`,
```
Comment (enahum) [inline store/sql_reaction_store.go]: Correct me if im wrong, but if this table has as PK `UserId`, `PostId` and `EmojiName` this check is redundant as the insert should fail with a duplicate key error

File: store/sql_upgrade.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ import (
 )
 
 const (
+	VERSION_3_6_0 = "3.6.0"
```
Comment (enahum) [inline store/sql_upgrade.go]: This should be in a separate PR

File: webapp/components/create_comment.jsx [hunk 3]
```diff
@@ -121,15 +119,43 @@ export default class CreateComment extends React.Component {
...
         }
 
-        MessageHistoryStore.storeMessageInHistory(this.state.message);
+        MessageHistoryStore.storeMessageInHistory(message);
+
+        if (message.trim().length === 0 && this.state.previews.length === 0) {
+            return;
+        }
+
+        const isReaction = (/^(\+|-):([^:\s]+):\s*$/).exec(message);
```
Comment (enahum) [inline webapp/components/create_comment.jsx]: can we have this regex somewhere common to try and avoid not having it update somewhere in case we need to?

File: webapp/components/create_post.jsx [hunk 4]
```diff
@@ -101,6 +103,7 @@ export default class CreatePost extends React.Component {
 
         this.setState({submitting: true, serverError: null});
 
+        const isReaction = (/^(\+|-):([^:\s]+):\s*$/).exec(post.message);
```
Comment (enahum) [inline webapp/components/create_post.jsx]: same here about the regex

File: store/sql_reaction_store.go (hunk 1 above)
Comment (enahum) [inline store/sql_reaction_store.go]: Or you can just check the error when inserting and return nil if it is duplicate key. I think it is a query that can be avoided

File: store/sql_upgrade.go (hunk 2 above)
Comment (hmhealey) [inline store/sql_upgrade.go]: I can submit it in a separate PR if we want to start doing that, but we [haven't done that in the past](https://github.com/mattermost/platform/commit/1e7985a87a72bea9a308cf1506dacc828c6e2e1c#diff-a788391365ece46cdf5c98badcdc3c32R18)

File: webapp/components/create_comment.jsx (hunk 3 above)
Comment (hmhealey) [inline webapp/components/create_comment.jsx]: Yep, I'll do that

File: api/reaction.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	if len(channelId) != 26 {
+		c.SetInvalidParam("deleteReaction", "postId")
+		return
+	}
+
+	if !HasPermissionToChannelContext(c, channelId, model.PERMISSION_READ_CHANNEL) {
+		return
+	}
+
+	postId := mux.Vars(r)["post_id"]
```
Comment (jwilander) [inline api/reaction.go]: Same comment here with the `mux.Vars(r)`

File: api/reaction.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		c.Err = result.Err
+		return
+	} else {
+		go sendReactionEvent(model.WEBSOCKET_EVENT_REACTION_REMOVED, channelId, reaction, postHadReactions)
+
+		ReturnStatusOK(w)
+	}
+}
+
+func sendReactionEvent(event string, channelId string, reaction *model.Reaction, postHadReactions bool) {
```
Comment (jwilander) [inline api/reaction.go]: 0/5 but everywhere this is used you already have the post, right? Maybe worth passing it in here instead of fetching it again

File: store/sql_post_store.go [hunk 7]
```diff
@@ -33,6 +33,7 @@ func NewSqlPostStore(sqlStore *SqlStore) PostStore {
 		table.ColMap("Props").SetMaxSize(8000)
 		table.ColMap("Filenames").SetMaxSize(4000)
 		table.ColMap("FileIds").SetMaxSize(150)
+		table.ColMap("HasReactions").SetMaxSize(1)
```
Comment (jwilander) [inline store/sql_post_store.go]: I don't think this is needed for boolean types

File: store/sql_reaction_store.go [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+			Posts
+		SET
+			UpdateAt = (CASE
+				WHEN HasReactions != (SELECT count(0) > 0 FROM Reactions WHERE PostId = :PostId) THEN :UpdateAt
+				ELSE UpdateAt
+			END),
+			HasReactions = (SELECT count(0) > 0 FROM Reactions WHERE PostId = :PostId)
+		WHERE
+			Id = :PostId`
+)
```
Comment (jwilander) [inline store/sql_reaction_store.go]: I'm 0/5 on this as well, but I'm not sure this exactly fits the spirit of constants (at least the way we use them) even if it does technically fit.

I think if you don't written out twice, maybe assign it to a variable closer to the functions it's actually used in?

File: store/sql_reaction_store.go [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+	return updatePostForReactions(transaction, reaction.PostId)
+}
+
+func updatePostForReactions(transaction *gorp.Transaction, postId string) error {
+	_, err := transaction.Exec(UPDATE_POST_HAS_REACTIONS_QUERY, map[string]interface{}{"PostId": postId, "UpdateAt": model.GetMillis()})
+
+	return err
+}
+
+func (s SqlReactionStore) List(postId string) StoreChannel {
```
Comment (jwilander) [inline store/sql_reaction_store.go]: 0/5 again but I think we call these `Get` everywhere else in store, maybe stick with that?

File: webapp/components/post_view/components/reaction.jsx [hunk 10]
```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2015 Mattermost, Inc. All Rights Reserved.
```
Comment (jwilander) [inline webapp/components/post_view/components/reaction.jsx]: 2016 again

File: webapp/components/post_view/components/reaction_list_container.jsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2015 Mattermost, Inc. All Rights Reserved.
```
Comment (jwilander) [inline webapp/components/post_view/components/reaction_list_container.jsx]: 2016 here too

File: webapp/components/post_view/components/reaction_list_view.jsx [hunk 12]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2015 Mattermost, Inc. All Rights Reserved.
```
Comment (jwilander) [inline webapp/components/post_view/components/reaction_list_view.jsx]: And here :P

File: api/reaction.go (hunk 6 above)
Comment (hmhealey) [inline api/reaction.go]: We have the post, but it's possibly outdated since `post.HasReactions` may have changed when the reaction was added/removed, so we still need to get it from the DB to check that. I could change the store call to return both the saved reaction and the updated post, but that would make it work differently than every other DB call in the app.

Comment (jwilander) [inline api/reaction.go]: Ok, makes sense. I think this way is fine

File: store/sql_reaction_store.go (hunk 8 above)
Comment (hmhealey) [inline store/sql_reaction_store.go]: I'll move it down to where its used, but I think that's the best we can do now.

The only reason it's used in multiple places is because one calls `Exec` with a transaction and the other doesn't. If I run into this again, I'll reconsider wrapping both the `DbMap` and `Transaction` types in a single interface since they have identical `Exec` calls. That way, we won't have to repeat query strings even in cases like this.

**Comment by hmhealey** [comment]
Reopening to trigger build

**Comment by alfredbez** [comment]
What happened to point 2 from the [spec](https://docs.google.com/document/d/1n7KGfMaf8EnRG3gYw2_bxf4Rzv4fzdyDzgKgll650xw/edit#) ? 

image

**Comment by jasonblais** [comment]
@alfredbez We didn't have an emoji picker yet, so didn't add point 2 from the spec.

The emoji picker is in progress (see #5157) after which we'll add the ability to add emoji reactions to any post.
